contracts/cosmos/verifier/verifier-constantine/lib/nim/verify/verify.nim

```diff
@@ -0,0 +1,89 @@
+else:
+  {.pragma: wasmPragma, cdecl, exportc, dynlib.}
+import
+  constantine/math/pairings/pairings_bn,
+  constantine/math/elliptic/[ec_shortweierstrass_affine, ec_shortweierstrass_projective],
+  constantine/math/io/[io_fields, io_bigints],
+  constantine/math/elliptic/ec_scalar_mul,
+  constantine/math/config/type_bigint,
+  constantine/math/arithmetic,
+  constantine/math/config/curves,
+  constantine/math/io/io_extfields
+
+import
+    verify_helpers
+
+type
+  IC* = array[3, ECP_ShortW_Aff[Fp[BN254_Snarks], G1]]
+
+  VerificationKey* = object
+    alpha*: ECP_ShortW_Aff[Fp[BN254_Snarks], G1]
+    beta*, gamma*, delta*: ECP_ShortW_Aff[Fp2[BN254_Snarks], G2]
+    ic*: IC
+
+  Proof* = object
+    a*, c*: ECP_ShortW_Aff[Fp[BN254_Snarks], G1]
+    b*: ECP_ShortW_Aff[Fp2[BN254_Snarks], G2]
+
+  Header* = object
+    head*: Fr[BN254_Snarks]
+    tail*: Fr[BN254_Snarks]
+
+  Input* = object
+    data*: ECP_ShortW_Aff[Fp[BN254_Snarks], G1]
+
+
+proc makePairsAndVerify*(vk: VerificationKey,
+                         prf: Proof,
+                         currentHeaderHash: var array[32, byte],
+                         newOptimisticHeader: var array[32, byte],
+                         newFinalizedHeader: array[32, byte],
+                         newExecutionStateRoot: array[32, byte],
+                         currentSlot: array[8, byte],
+                         domain: array[32, byte]): bool {.wasmPragma.} =
+  var reverseSlot: array[8, byte]
+  for i in 0..7:
+    reverseSlot[i] = currentSlot[7-i]
+  var zerosSlotBuffer: array[24, byte]
+  for i in 0..23:
+    zerosSlotBuffer[i] = 0
+
+  let sha256ofHashes = hashHeaders(currentHeaderHash, newOptimisticHeader,
+                                   newFinalizedHeader, newExecutionStateRoot,
+                                   zerosSlotBuffer, reverseSlot, domain)
+  let header = headerFromSeq(@sha256ofHashes)
+
+  var preparedInputs:Input
+  var ic0Prj: ECP_ShortW_Prj[Fp[BN254_Snarks], G1]
+  var ic1Prj: ECP_ShortW_Prj[Fp[BN254_Snarks], G1]
+  var ic2Prj: ECP_ShortW_Prj[Fp[BN254_Snarks], G1]
+  fromAffine(ic0Prj, vk.ic[0])
+  fromAffine(ic1Prj, vk.ic[1])
+  fromAffine(ic2Prj, vk.ic[2])
+  scalarMul(ic1Prj,toBig(header.head))
+  scalarMul(ic2Prj,toBig(header.tail))
+  ic0Prj += ic1Prj
+  ic0Prj += ic2Prj
+  affine(preparedInputs.data,ic0Prj)
+
+  var aBPairing: Fp12[BN254_Snarks]
+  pairing_bn[BN254_Snarks](aBPairing,prf.a,prf.b)
+  var alphaBetaPairing: Fp12[BN254_Snarks]
+  pairing_bn[BN254_Snarks](alphaBetaPairing,vk.alpha,vk.beta)
+  var preparedInputsGammaPairing: Fp12[BN254_Snarks]
+  pairing_bn[BN254_Snarks](preparedInputsGammaPairing,preparedInputs.data,vk.gamma)
+  var proofCVkDeltaPairing: Fp12[BN254_Snarks]
+  pairing_bn[BN254_Snarks](proofCVkDeltaPairing,prf.c,vk.delta)
+
+  var sum:Fp12[BN254_Snarks]
+  prod(sum, alphaBetaPairing, preparedInputsGammaPairing)
+  prod(sum, sum, proofCVkDeltaPairing)
+
+  if (sum == aBPairing).bool:
```

Comment:
It would be almost 2x faster if multipairings were used there and the result checked for one:

with the signature
```Nim
func pairing*[N: static int, C](
       gt: var Fp12[C],
       Ps: array[N, ECP_ShortW_Aff[Fp[C], G1]],
       Qs: array[N, ECP_ShortW_Aff[Fp2[C], G2]]) =
```

from https://github.com/mratsim/constantine/blob/d69c7bf/constantine/math/pairings/pairings_generic.nim#L16C1-L22
forwarding to https://github.com/mratsim/constantine/blob/d69c7bf/constantine/math/pairings/pairings_bn.nim#L178-L181

`prf.a` needs to be negated beforehand because:

`e(prf.a, prf.b) = e(vk.alpha, vk.beta).e(preparedInputs.data, vk.gamma) = e(prf.c,vk.delta)`

is equivalent to

`e(-prf.a, prf.b).e(vk.alpha, vk.beta).e(preparedInputs.data, vk.gamma) = e(prf.c,vk.delta) = 1`

see BLS signatures: https://github.com/mratsim/constantine/blob/d69c7bf/constantine/signatures/bls_signatures.nim#L123-L127